## LAB 9 - SVM FOR MULTICLASS CLASSIFICATION (24.02.2025)

## AIM
To apply SVM classifier to the Wine Dataset using **One versus One** and **One Versus Rest** evaluate the model's perfomance using `classification_report`. 

## ALGORITHM

1. **Load Data:**
   - Load the Wine dataset (features `X` and target labels `y`).
   - Visualise the dataset using `DataFrame`

2. **Split Data:**
   - Split the dataset into training and testing sets (e.g., 70% training, 30% testing).

3. **Normalize Data:**
   - Apply `StandardScaler` to normalize the training and testing data.

4. **Train SVC:**
   - Create two `SVC` model with appropriate parameters (`gamma` and `C`) to avoid overfitting.
   - The first `SVC` model will have `decision_function_shape` as `ovo` (One-versus-One)
   - The second `SVC` model will have `decision_function_shape` as `ovr` (One-versus-Rest)
   - Fit the models on the training data.

5. **Make Predictions:**
   - Predict labels for the test set using the trained model.

6. **Evaluate Model:**
   - Print classification report: `classification_report(y_test, y_pred)` for both the models, from which evaluation metrics like precision,recall, and accuracy are returned.

## DESCRIPTION

Support Vector Machines (SVM) for multiclass classification typically use strategies like **one-vs-one (OvO)** or **one-vs-all (OvA)**. 

- **One-vs-all (OvA)**: A separate binary SVM classifier is trained for each class, where the goal is to separate that class from all other classes. The class with the highest confidence score for a given test point is predicted.
  
- **One-vs-one (OvO)**: A binary SVM classifier is trained for every pair of classes. For \(k\) classes, this results in \(k(k-1)/2\) classifiers. The final prediction is made based on voting from all classifiers.

SVMs aim to find a hyperplane that maximizes the margin between classes, and in multiclass classification, this is extended to separate multiple classes.

## CODE AND OUTPUT

In [35]:
import pandas as pd
import numpy as np
from sklearn.datasets import load_wine

In [38]:
df = pd.DataFrame(load_wine()['data'],columns=load_wine()['feature_names'])
df['target'] = load_wine()['target']
df.head()

alcohol  malic_acid   ash  alcalinity_of_ash  magnesium  total_phenols  \
0    14.23        1.71  2.43               15.6      127.0           2.80   
1    13.20        1.78  2.14               11.2      100.0           2.65   
2    13.16        2.36  2.67               18.6      101.0           2.80   
3    14.37        1.95  2.50               16.8      113.0           3.85   
4    13.24        2.59  2.87               21.0      118.0           2.80   

   flavanoids  nonflavanoid_phenols  proanthocyanins  color_intensity   hue  \
0        3.06                  0.28             2.29             5.64  1.04   
1        2.76                  0.26             1.28             4.38  1.05   
2        3.24                  0.30             2.81             5.68  1.03   
3        3.49                  0.24             2.18             7.80  0.86   
4        2.69                  0.39             1.82             4.32  1.04   

   od280/od315_of_diluted_wines  proline  target  
0                          3.92   1065.0       0  
1                          3.40   1050.0       0  
2                          3.17   1185.0       0  
3                          3.45   1480.0       0  
4                          2.93    735.0       0

In [26]:
X,y = load_wine(return_X_y=True)

In [27]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X,y,random_state=42,test_size=0.3)

In [28]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

In [29]:
from sklearn.svm import SVC
svm_ovo = SVC(decision_function_shape='ovo',kernel="rbf", gamma='auto', C=1.0)
svm_ovo.fit(X_train,y_train)

SVC(decision_function_shape='ovo', gamma='auto')

In [30]:
from sklearn.metrics import classification_report
y_pred_ovo = svm_ovo.predict(X_test)

In [31]:
print(classification_report(y_test,y_pred_ovo))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        19
           1       0.95      1.00      0.98        21
           2       1.00      0.93      0.96        14

    accuracy                           0.98        54
   macro avg       0.98      0.98      0.98        54
weighted avg       0.98      0.98      0.98        54



In [33]:
y_pred_ovr = svm_ovr.predict(X_test)

In [34]:
print(classification_report(y_test,y_pred_ovo))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        19
           1       0.95      1.00      0.98        21
           2       1.00      0.93      0.96        14

    accuracy                           0.98        54
   macro avg       0.98      0.98      0.98        54
weighted avg       0.98      0.98      0.98        54



## RESULT

Two SVM classifiers has been trained on the wine dataset with an accuracy of 98% and 98%.